### PR TITLE
[GH-125] Remember user last status on status change

### LINKED ISSUE #125: Do not automatically update user's status from Offline
There are a couple design gaps with the current implementation of status change. When the status is set to change without user confirmation there are issues.

1. ~~When confirm status change is set to `No` and the user is `offline` we should not change the users status. There does not need to be any change in behavior when confirm status change is set to `Yes`.~~ Fixed by  @larkox in a separate PR
2. When a users status is updated by the job there are cases where we should store the previous status and return the user to that status rather than `online`

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
The plugin will remember the last status before a meeting and get back to that status after the meeting if the status is assigned manually. If not, it will default to Online.

#### Ticket Link
Fix #125

### CODE REVIEW

File: server/mscalendar/views/status_change_notification.go [hunk 1]
```diff
@@ -49,7 +49,7 @@ func RenderEventWillStartLine(subject, weblink string, startTime time.Time) stri
 
 func renderScheduleItem(event *remote.Event, status string) string {
 	if event == nil {
-		return fmt.Sprintf("You have no upcoming events.\n Shall I change your status to %s?", prettyStatuses[status])
+		return fmt.Sprintf("You have no upcoming events.\n Shall I change back your status to %s?", prettyStatuses[status])
```
Comment (mickmister) [inline server/mscalendar/views/status_change_notification.go:52]: 0/5
```suggestion
		return fmt.Sprintf("You have no upcoming events.\n Shall I change your status back to %s?", prettyStatuses[status])
```

File: server/store/user_store.go [hunk 2]
```diff
@@ -38,8 +38,9 @@ type User struct {
 	Remote            *remote.User
 	MattermostUserID  string
 	OAuth2Token       *oauth2.Token
-	Settings          Settings          `json:"mattermostSettings,omitempty"`
-	ActiveEvents      []string          `json:"events"`
+	Settings          Settings `json:"mattermostSettings,omitempty"`
+	ActiveEvents      []string `json:"events"`
+	LastStatus        string
```
Comment (mickmister) [inline server/store/user_store.go:43]: Do you see any value in storing this as a `*model.Status` instead? Currently we don't benefit from it based on how the code is written but I'm just wondering.

File: server/mscalendar/availability.go [hunk 3]
```diff
@@ -276,19 +285,40 @@ func (m *mscalendar) setStatusFromCalendarView(user *store.User, currentStatus s
...
+		}
+	}
+
+	err := m.Store.StoreUser(user)
+	if err != nil {
+		return err
+	}
+
+	if !user.Settings.GetConfirmation {
+		_, appErr := m.PluginAPI.UpdateMattermostUserStatus(user.MattermostUserID, toSet)
```
Comment (mickmister) [inline server/mscalendar/availability.go:311]: So if the user has asked to get confirmation changes, they also get asked after the meeting? That makes sense, as then we are not ever changing the user's status without their approval.

@asaadmahmood @aaronrothschild Thoughts on this?

File: server/mscalendar/availability.go [hunk 4]
```diff
@@ -276,19 +285,40 @@ func (m *mscalendar) setStatusFromCalendarView(user *store.User, currentStatus s
 	return message, nil
 }
 
-func (m *mscalendar) setStatusOrAskUser(user *store.User, status string, events []*remote.Event) error {
-	if user.Settings.GetConfirmation {
-		url := fmt.Sprintf("%s%s%s", m.Config.PluginURLPath, config.PathPostAction, config.PathConfirmStatusChange)
-		_, err := m.Poster.DMWithAttachments(user.MattermostUserID, views.RenderStatusChangeNotificationView(events, status, url))
-		if err != nil {
-			return err
+func (m *mscalendar) setStatusOrAskUser(user *store.User, status *model.Status, events []*remote.Event, isFree bool) error {
```
Comment (mickmister) [inline server/mscalendar/availability.go:288]:
```suggestion
func (m *mscalendar) setStatusOrAskUser(user *store.User, currentStatus *model.Status, events []*remote.Event, isFree bool) error {
```

The parameters to this function seem sort of cryptic. The fact that we need to access `user` to determine their previous status, and that it is a fundamental piece of the logic here, is a bit confusing.

For example, I would expect the `status` parameter to be the status we are setting the user to, but that parameter is actually the current status. We already know what status we need to set the user to in the previous function right? I suppose it's fine that this function is performing the logic, but it's a bit hard to read. Maybe put some comments in this function?

**Comment by mickmister** [comment]
@larkox Looks like status change tests are failing btw

(Also I accidentally clicked the close PR button. Why isn't there a confirmation for that 🙃)

File: server/mscalendar/views/status_change_notification.go (hunk 1 above)
Comment (larkox) [reply server/mscalendar/views/status_change_notification.go:52]: You are the native speaker, so I will go with your suggestion.

File: server/store/user_store.go (hunk 2 above)
Comment (larkox) [reply server/store/user_store.go:43]: The only "useful" information for us in model.Status as far as I can see is the Manual bool, but we don't need it after we have decided that it was a manual status. That being said, we could consider storing that if we would like to do some more complex handling in the future (store the status and whether it was manual or not) but I think it is not worth it.

File: server/mscalendar/availability.go (hunk 3 above)
Comment (larkox) [reply server/mscalendar/availability.go:311]: I thought this was like this by design from the beginning. As far as I know, it has always been implemented this way.

**Comment by larkox** [comment]
@mickmister comments addressed.
Regarding testing, the current tests are starting to become a bit of outdated and hard to follow. I would create a new ticket to refactor those tests with all the new information and more test cases. Maybe a HW ticket?

File: server/mscalendar/availability.go [hunk 5]
```diff
@@ -285,7 +285,12 @@ func (m *mscalendar) setStatusFromCalendarView(user *store.User, status *model.S
 	return message, nil
 }
 
-func (m *mscalendar) setStatusOrAskUser(user *store.User, status *model.Status, events []*remote.Event, isFree bool) error {
+// setStatusOrAskUser to which status change, and whether it should update the status automatically or ask the user.
+// - user: the user to change the status
```
Comment (mickmister) [inline server/mscalendar/availability.go:289]:
```suggestion
// - user: the user to change the status. We use user.LastStatus to determine the status the user had before the beginning of the meeting.
```

File: server/mscalendar/availability.go [hunk 6]
```diff
@@ -285,7 +285,12 @@ func (m *mscalendar) setStatusFromCalendarView(user *store.User, status *model.S
...
+// - currentStatus: currentStatus, to decide whether to store this status when the user is free
```
Comment (mickmister) [inline server/mscalendar/availability.go:290]:
```suggestion
// - currentStatus: currentStatus, to decide whether to store this status when the user is free. This gets assigned to user.LastStatus at the beginning of the meeting.
```

**Comment by DHaussermann** [comment]
@larkox I have tried to do another test pass on this but the plugin is crashing. Unfortunately it seems this PR has gone stale and needs to be synced with master. 

Logs show:
```
{"level":"debug","ts":1599082411.078088,"caller":"plugin/health_check.go:72","msg":"Restarting plugin due to failed health check","id":"com.mattermost.mscalendar"}
{"level":"error","ts":1599082411.078157,"caller":"mlog/log.go:190","msg":"RPC call OnDeactivate to plugin failed.","plugin_id":"com.mattermost.mscalendar","error":"connection is shut down"}
{"level":"warn","ts":1599082411.078331,"caller":"plugin/hclog_adapter.go:69","msg":"error closing client during Kill","plugin_id":"com.mattermost.mscalendar","wrapped_extras":"errconnection is shut down"}
{"level":"warn","ts":1599082411.078377,"caller":"plugin/hclog_adapter.go:71","msg":"plugin failed to exit gracefully","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.097565,"caller":"plugin/hclog_adapter.go:51","msg":"starting plugin","plugin_id":"com.mattermost.mscalendar","wrapped_extras":"pathplugins/com.mattermost.mscalendar/server/dist/plugin-darwin-amd64args[plugins/com.mattermost.mscalendar/server/dist/plugin-darwin-amd64]"}
{"level":"debug","ts":1599082411.097522,"caller":"web/handlers.go:100","msg":"Received HTTP request","method":"GET","url":"/api/v4/teams/mkxtxnec8trmu8iiiygeucmfxo/commands/autocomplete_suggestions","request_id":"3if95yfsstgotp338exm74jt4r","status_code":"200"}
{"level":"debug","ts":1599082411.10598,"caller":"plugin/hclog_adapter.go:51","msg":"plugin started","plugin_id":"com.mattermost.mscalendar","wrapped_extras":"pathplugins/com.mattermost.mscalendar/server/dist/plugin-darwin-amd64pid19543"}
{"level":"debug","ts":1599082411.1061099,"caller":"plugin/hclog_adapter.go:51","msg":"waiting for RPC address","plugin_id":"com.mattermost.mscalendar","wrapped_extras":"pathplugins/com.mattermost.mscalendar/server/dist/plugin-darwin-amd64"}
{"level":"debug","ts":1599082411.123156,"caller":"plugin/hclog_adapter.go:51","msg":"plugin address","plugin_id":"com.mattermost.mscalendar","wrapped_extras":"address/var/folders/zq/326_nty97s325f52_1r8r8qc0000gn/T/plugin880101250networkunixtimestamp2020-09-02T17:33:31.122-0400"}
{"level":"debug","ts":1599082411.123064,"caller":"plugin/hclog_adapter.go:51","msg":"using plugin","plugin_id":"com.mattermost.mscalendar","wrapped_extras":"version1"}
{"level":"debug","ts":1599082411.219154,"caller":"plugin/hclog_adapter.go:53","msg":"panic: runtime error: invalid memory address or nil pointer dereference","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.219219,"caller":"plugin/hclog_adapter.go:53","msg":"[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1a17fdf]","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.219241,"caller":"plugin/hclog_adapter.go:53","msg":"","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.219258,"caller":"plugin/hclog_adapter.go:53","msg":"goroutine 26 [running]:","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.2192729,"caller":"plugin/hclog_adapter.go:53","msg":"github.com/mattermost/mattermost-plugin-mscalendar/server/plugin.(*Plugin).OnConfigurationChange.func2(0xc000123618)","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.219305,"caller":"plugin/hclog_adapter.go:53","msg":"\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-mscalendar/server/plugin/plugin.go:158 +0x51f","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.219844,"caller":"plugin/hclog_adapter.go:53","msg":"github.com/mattermost/mattermost-plugin-mscalendar/server/plugin.(*Plugin).updateEnv(0xc0001235f0, 0xc0001efa78, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.219885,"caller":"plugin/hclog_adapter.go:53","msg":"\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-mscalendar/server/plugin/plugin.go:258 +0x95","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.2199051,"caller":"plugin/hclog_adapter.go:53","msg":"github.com/mattermost/mattermost-plugin-mscalendar/server/plugin.(*Plugin).OnConfigurationChange(0xc0001235f0, 0x0, 0x0)","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.219924,"caller":"plugin/hclog_adapter.go:53","msg":"\t/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-mscalendar/server/plugin/plugin.go:142 +0x426","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.2199428,"caller":"plugin/hclog_adapter.go:53","msg":"github.com/mattermost/mattermost-server/v5/plugin.(*hooksRPCServer).OnActivate(0xc000199de0, 0xc00018b638, 0xc00026bb10, 0x0, 0x0)","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.2199612,"caller":"plugin/hclog_adapter.go:53","msg":"\t/Users/dylanhaussermann/go/pkg/mod/github.com/mattermost/mattermost-server/v5@v5.3.2-0.20200529093641-c9cdeba1a7e2/plugin/client_rpc.go:216 +0x225","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.219979,"caller":"plugin/hclog_adapter.go:53","msg":"reflect.Value.call(0xc0003bb140, 0xc0000103b0, 0x13, 0x1dd7d95, 0x4, 0xc00006af08, 0x3, 0x3, 0xc000059f34, 0xcee53202140bc, ...)","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.219999,"caller":"plugin/hclog_adapter.go:53","msg":"\t/usr/local/go/src/reflect/value.go:460 +0x8ab","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.2200162,"caller":"plugin/hclog_adapter.go:53","msg":"reflect.Value.Call(0xc0003bb140, 0xc0000103b0, 0x13, 0xc000059f08, 0x3, 0x3, 0xcee53202140bc, 0xc0000c6070, 0xc0000c6070)","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.220032,"caller":"plugin/hclog_adapter.go:53","msg":"\t/usr/local/go/src/reflect/value.go:321 +0xb4","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.220048,"caller":"plugin/hclog_adapter.go:53","msg":"net/rpc.(*service).call(0xc000043dc0, 0xc00007ed20, 0xc0000394d0, 0xc0000394e0, 0xc000119b80, 0xc0000a05e0, 0x1ac0be0, 0xc00018b638, 0x16, 0x1ac0c20, ...)","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.220065,"caller":"plugin/hclog_adapter.go:53","msg":"\t/usr/local/go/src/net/rpc/server.go:377 +0x17f","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.2200809,"caller":"plugin/hclog_adapter.go:53","msg":"created by net/rpc.(*Server).ServeCodec","plugin_id":"com.mattermost.mscalendar"}
{"level":"debug","ts":1599082411.220097,"caller":"plugin/hclog_adapter.go:53","msg":"\t/usr/local/go/src/net/rpc/server.go:474 +0x42b","plugin_id":"com.mattermost.mscalendar"}
{"level":"error","ts":1599082411.220684,"caller":"mlog/log.go:190","msg":"RPC call to OnActivate plugin failed.","plugin_id":"com.mattermost.mscalendar","error":"unexpected EOF"}
```

cc @mickmister  in case you're able to merge master in as I believe Daniel is away.

**Comment by mickmister** [comment]
@DHaussermann This is happening due to the `EnableDiagnostics` flag being false. I'll create a separate PR to fix that. For now you can enable diagnostics. But I do like the fact that you've been able to catch these.

**Comment by DHaussermann** [comment]
@larkox  I have done another round of testing and ran through various scenarios for cases with DND vs Away and status change notification on vs off 

There is 1 scenario where at the end of the meeting the user's status does not seem to return as expected. Are you able to reproduce this? 

Steps:
1. Set status as Away during meeting events 
2. Set status change confirmation to Yes
3. Set user's status to Online
4.  When meeting approaches - Click Yes and confirm flipping status to Away
5. Wait for availability job to run after even has ended

**Observed**:    
Even though user is already set to Away during the meeting - After the meeting user receive a notification requesting confirmation they want their status changes to Away as they are not in a meting
SetBackToAway
SetBackToAway2

**Expected**:
User is asked if they would like their status changed to Online

**Comment by larkox** [comment]
@DHaussermann I think it was a problem of the last status not being "cleared". Let me know if you keep seen the same error.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Various cases with setting "busy" status as Away and DND
- These cases where done for all different statuses as the original status (Away, DND, Online and Offline)
- These cases where done both with and without status change notification enabled
- Tested that status revert to previous status before the meeting does not occur if the user has adjusted their status manually during the meeting 
- Issue from last build where you do not correctly return from Away has been resolved
LGTM!

Test cases documented in release testing. I will move them into TM4J soon.
